packages/teuchos/comm/src/Teuchos_StackedTimer.cpp

```diff
@@ -0,0 +1,204 @@
+// @HEADER BEGIN
+// @HEADER END
+
+#include "Teuchos_StackedTimer.hpp"
+#include <limits>
+
+
+namespace Teuchos {
+
+Teuchos::RCP<StackedTimer> StackedTimer::timer;
+  
+StackedTimer::LevelTimer::LevelTimer() :
+    level_(std::numeric_limits<unsigned>::max()),name_("INVALID"),parent_(NULL)
+{}
+
+void error_out(const std::string& msg, const bool)
+{
+  TEUCHOS_TEST_FOR_EXCEPTION(true,std::runtime_error,msg);
+}
+  
+
+void
+StackedTimer::LevelTimer::report(std::ostream &os) {
+  for (unsigned i=0; i<level_; ++i)
+    os << "   ";
+  os << name_<<":"<<accumulatedTime()<< " [" << count_started_<<"] ("<< count_updates_ <<")"<<std::endl;
+  double t_total = 0;
+  for (size_t i=0; i<sub_timers_.size(); ++i) {
+    t_total += sub_timers_[i].accumulatedTime();
+    sub_timers_[i].report(os);
+  }
+  if ( sub_timers_.size() == 0 )
+    return;
+  for (unsigned i=0; i<=level_; ++i)
+    os << "   ";
+  os << "Remainder: " << accumulatedTime() - t_total<<std::endl;
+
+}
+
+BaseTimer::TimeInfo
+StackedTimer::LevelTimer::findTimer(std::string &name) {
+  BaseTimer::TimeInfo t;
+  if (get_full_name() == name)
+    t = BaseTimer::TimeInfo(this);
+  else {
+    for (unsigned i=0;i<sub_timers_.size(); ++i){
+      t = sub_timers_[i].findTimer(name);
+      if (t.time > 0.)
+        return t;
+    }
+  }
+  return t;
+}
+void
+StackedTimer::flatten() {
+  int num_timers = timer_.countTimers();
+  flat_names_.resize(num_timers);
+  unsigned pos=0;
+  timer_.addTimerNames(flat_names_, pos);
+}
+
+void
+StackedTimer::merge(Teuchos::RCP<const Teuchos::Comm<int> > comm){
+  Array<std::string> all_names;
+  mergeCounterNames(*comm, flat_names_, all_names, Union);
+  flat_names_ = all_names;
+}
+
+void
+StackedTimer::collectRemoteData(Teuchos::RCP<const Teuchos::Comm<int> > comm) {
+  int comm_size = size(*comm);
+  int comm_rank = rank(*comm);
+  Array<BaseTimer::TimeInfo> local_time_data;
+  if (comm_rank == 0) {
+    time_data_.resize(comm_size);
+    for (int r=0; r<comm_size; ++r)
+      time_data_[r].resize(flat_names_.size());
+  } else {
+    local_time_data.resize(flat_names_.size());
+  }
+
+  for (unsigned i=0; i<flat_names_.size(); ++i) {
+    auto t = timer_.findTimer(flat_names_[i]);
+    if ( comm_rank == 0 )
+      time_data_[0][i] = t;
+    else
+      local_time_data[i] = t;
+  }
+
+  int buffer_size = sizeof(BaseTimer::TimeInfo[flat_names_.size()]);
+
+  if (comm_rank == 0 )  {
```

Comment:
No love for `MPI_Allgather` and `MPI_Allgatherv`?

Reply:
Moreover, `allgather` is an inherently non-scalable operation. If we are just gathering statistics, there are scalable ways of doing this with just reductions.